iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp

```diff
@@ -734,7 +742,77 @@ class RankReducedDotGeneral : public OpRewritePattern<mhlo::DotGeneralOp> {
 
     return success();
   }
-};  // namespace
+};
+
+// Generates Gaussian noise with uniform random generator based on Box-Muller
+// transform.
+class ExpandRngNormalToRngUniform : public OpRewritePattern<mhlo::RngNormalOp> {
+ public:
+  using OpRewritePattern<mhlo::RngNormalOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::RngNormalOp op,
+                                PatternRewriter &rewriter) const override {
+    auto resTy = op.getType().dyn_cast<RankedTensorType>();
+    // We can support static shapes, but it's easier to implement Box-Muller
+    // transform if we know the number of elements.
+    if (!resTy || !resTy.hasStaticShape()) return failure();
+
+    // The algorithm requires even numbers and will generate pairs.
+    auto numElems = resTy.getNumElements();
+    if (numElems & 1) numElems++;
+    auto halfNumElems = numElems / 2;
+
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+    Value epsilon = getF32SplatConst(b, /*shapes=*/{},
+                                     std::numeric_limits<float>::epsilon());
+    Value upperBound = getF32SplatConst(b, /*shapes=*/{}, 1.0);
+
+    // Create pairs of random numbers, make sure u1 is greater than epsilon.
+    auto shapeTensor =
+        b.create<mhlo::ConstOp>(make1DElementsAttr(rewriter, numElems));
+    auto rngType = RankedTensorType::get({numElems}, resTy.getElementType());
+    auto rng =
+        b.create<mhlo::RngUniformOp>(rngType, epsilon, upperBound, shapeTensor);
+    OpFoldResult zero = b.getIndexAttr(0);
+    OpFoldResult one = b.getIndexAttr(1);
+    OpFoldResult half = b.getIndexAttr(halfNumElems);
+    auto u1 = b.create<tensor::ExtractSliceOp>(rng, zero, half, one);
+    auto u2 = b.create<tensor::ExtractSliceOp>(rng, half, half, one);
+
+    // mag = sigma * sqrt(-2.0 * log(u1));
+    Value sigma = b.create<mhlo::BroadcastOp>(
+        u1.getType(), op.sigma(), make1DElementsAttr(b, halfNumElems));
+    Value mag = b.create<mhlo::LogOp>(u1);
+    mag = b.create<mhlo::MulOp>(
+        mag, getF32SplatConst(b, /*shapes=*/{halfNumElems}, -2.0));
+    mag = b.create<mhlo::MulOp>(sigma, b.create<mhlo::SqrtOp>(mag));
+
+    // z0 = mag * cos(two_pi * u2) + mu;
+    // z1 = mag * sin(two_pi * u2) + mu;
+    Value twoPi = getF32SplatConst(b, /*shapes=*/{halfNumElems}, 2.0 * M_PI);
+    Value mu = b.create<mhlo::BroadcastOp>(u1.getType(), op.mu(),
```

Comment:
Nit: Since we are running this at runtime, It might be faster to do do inverse transform sampling that doesn't require evaluating sine or cosine. Or revisit eliminating them (IIRC there is a way to do Box-Muller without any trigonometric function).

Reply:
Good point, I think we can eliminate trigonometric functions with polar form. I think we won't hit such issue because I'm going to  compute these at compile time.